In [5]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

In [6]:
raw_data = pd.read_csv('data.csv')
raw_data.head()

SAT Admitted  Gender
0  1363       No    Male
1  1792      Yes  Female
2  1954      Yes  Female
3  1653       No    Male
4  1593       No    Male

In [8]:
data = raw_data.copy()
data['Admitted'] = data['Admitted'].map({'Yes':1, 'No':0})
data['Gender'] = data['Gender'].map({'Female':1, 'Male': 0})
data.head()

SAT  Admitted  Gender
0  1363         0       0
1  1792         1       1
2  1954         1       1
3  1653         0       0
4  1593         0       0

In [14]:
y = data['Admitted']
x1 = data[['SAT', 'Gender']]

In [15]:
x = sm.add_constant(x1)
reg_log = sm.Logit(y,x)
results_log = reg_log.fit()
results_log.summary()

Optimization terminated successfully.
         Current function value: 0.120117
         Iterations 10


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:               Admitted   No. Observations:                  168
Model:                          Logit   Df Residuals:                      165
Method:                           MLE   Df Model:                            2
Date:                Tue, 06 Aug 2019   Pseudo R-squ.:                  0.8249
Time:                        23:17:09   Log-Likelihood:                -20.180
converged:                       True   LL-Null:                       -115.26
Covariance Type:            nonrobust   LLR p-value:                 5.118e-42
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const        -68.3489     16.454     -4.154      0.000    -100.598     -36.100
SAT            0.0406      0.010      4.129      0.000       0.021       0.060
Gender         1.9449      0.846      2.299      0.022       0.287       3.603
==============================================================================

Possibly complete quasi-separation: A fraction 0.27 of observations can be
perfectly predicted. This might indicate that there is complete
quasi-separation. In this case some parameters will not be identified.
"""

#### log(odds) = -0.64 + 2.08 * Gender

log(odds_2) = -68.3489 + 1.9449 * Gender_2  
log(odds_1) = -68.3489 + 1.9449 * Gender_1  
log(odds_2/odds_1) = 1.9449 * (Gender_2-Gender_1)  
If Gender_2 = female and Gender_1 = male:  
log(odds_female / odds_male) = 1.9449 * (1-0) = 1.9449  

In [17]:
np.exp(1.9449)

6.992932526814459

odds_female = 6.99 * odds_male  
Given the same SAT score, a female has 7 times higher odds to get admitted  
### Accuracy

In [23]:
np.set_printoptions(formatter={'float': lambda x: "{0:0.2f}".format(x)})
results_log.predict()
#Predicted Values By The Model

array([0.00, 1.00, 1.00, 0.23, 0.02, 0.99, 1.00, 1.00, 1.00, 0.01, 1.00,
       1.00, 0.76, 0.00, 0.60, 1.00, 0.11, 0.12, 0.51, 1.00, 1.00, 1.00,
       0.00, 0.01, 0.97, 1.00, 0.48, 0.99, 1.00, 0.99, 0.00, 0.83, 0.25,
       1.00, 1.00, 1.00, 0.31, 1.00, 0.23, 0.00, 0.02, 0.45, 1.00, 0.00,
       0.99, 0.00, 0.99, 0.00, 0.00, 0.01, 0.00, 1.00, 0.92, 0.02, 1.00,
       0.00, 0.37, 0.98, 0.12, 1.00, 0.00, 0.78, 1.00, 1.00, 0.98, 0.00,
       0.00, 0.00, 1.00, 0.00, 0.78, 0.12, 0.00, 0.99, 1.00, 1.00, 0.00,
       0.30, 1.00, 1.00, 0.00, 1.00, 1.00, 0.85, 1.00, 1.00, 0.00, 1.00,
       1.00, 0.89, 0.83, 0.00, 0.98, 0.97, 0.00, 1.00, 1.00, 0.03, 0.99,
       0.96, 1.00, 0.00, 1.00, 0.01, 0.01, 1.00, 1.00, 1.00, 0.00, 0.00,
       0.02, 0.33, 0.00, 1.00, 0.09, 0.00, 0.97, 0.00, 0.75, 1.00, 1.00,
       0.01, 0.01, 0.00, 1.00, 0.00, 0.99, 0.57, 0.54, 0.87, 0.83, 0.00,
       1.00, 0.00, 0.00, 0.00, 1.00, 0.04, 0.00, 0.01, 1.00, 0.99, 0.52,
       1.00, 1.00, 0.05, 0.00, 0.00, 0.00, 0.68, 1.

In [24]:
np.array(data['Admitted'])
#Actual Values

array([0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0,
       1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0,
       0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0,
       1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0], dtype=int64)

In [25]:
results_log.pred_table()

array([[69.00, 5.00],
       [4.00, 90.00]])

In [27]:
cm_df = pd.DataFrame(results_log.pred_table())
cm_df.columns = ['Predicted 0', 'Predicted 1']
cm_df = cm_df.rename(index={0:'Actual 0', 1:'Actual 1'})
cm_df
#Confusion Matrix

Predicted 0  Predicted 1
Actual 0         69.0          5.0
Actual 1          4.0         90.0

In [29]:
#we got accurate prediction in 159 out of 168 cases, 159/168 = 94.6% accuracy
cm = np.array(cm_df)
accuracy_train = (cm[0,0] + cm[1,1])/cm.sum()
accuracy_train

0.9464285714285714

## Testing and assessing it's accuracy

In [30]:
test = pd.read_csv('test.csv')
test.head()

SAT Admitted  Gender
0  1323       No    Male
1  1725      Yes  Female
2  1762      Yes  Female
3  1777      Yes    Male
4  1665       No    Male

In [31]:
test['Admitted'] = test['Admitted'].map({'Yes':1, 'No':0})
test['Gender'] = test['Gender'].map({'Female':1, 'Male': 0})
test.head()

SAT  Admitted  Gender
0  1323         0       0
1  1725         1       1
2  1762         1       1
3  1777         1       0
4  1665         0       0

In [32]:
x.head()

const   SAT  Gender
0    1.0  1363       0
1    1.0  1792       1
2    1.0  1954       1
3    1.0  1653       0
4    1.0  1593       0

In [33]:
test_actual = test['Admitted']
test_data = test.drop(['Admitted'], axis=1)
test_data = sm.add_constant(test_data)
test_data.head()

const   SAT  Gender
0    1.0  1323       0
1    1.0  1725       1
2    1.0  1762       1
3    1.0  1777       0
4    1.0  1665       0

In [36]:
def confusion_matrix(data, actual_values, model):
    
    pred_values = model.predict(data)
    bins = np.array([0, 0.5, 1])
    cm = np.histogram2d(actual_values, pred_values, bins=bins)[0]
    accuracy = (cm[0,0] + cm[1,1]) / cm.sum()
    return cm, accuracy

In [37]:
cm = confusion_matrix(test_data, test_actual, results_log)
cm

(array([[5.00, 1.00],
        [1.00, 12.00]]), 0.8947368421052632)

In [38]:
print("Missclassification rate: " + str((1+1)/19))

Missclassification rate: 0.10526315789473684
